File: backend-java/src/main/java/com/skewwhiffy/auraltester/notation/model/note/Octave.java
package com.skewwhiffy.auraltester.notation.model.note;

import com.skewwhiffy.auraltester.dao.OctaveDao;
import org.jetbrains.annotations.NotNull;

public record Octave(int offsetFromDefault) implements Comparable<Octave> {
    public OctaveDao toDao() {
        return new OctaveDao(offsetFromDefault);
    }

    public Octave up() {
        return new Octave(offsetFromDefault + 1);
    }

    public Octave down() {
        return new Octave(offsetFromDefault - 1);
    }

    @Override
    public int compareTo(@NotNull Octave o) {
        return offsetFromDefault - o.offsetFromDefault;
    }

    public static Octave getDefault() {
        return new Octave(0);
    }
}

File: backend.old/src/main/kotlin/com/skewwhiffy/auraltester/notation/model/note/Octave.kt
package com.skewwhiffy.auraltester.notation.model.note

import com.skewwhiffy.auraltester.dao.OctaveDao

// @JvmRecord
data class Octave(val offsetFromDefault: Int) : Comparable<Octave?> {
    companion object {
        val default: Octave
            get() = Octave(0)
    }

    override fun compareTo(other: Octave?): Int {
        if (other == null) {
            return 1
        }
        return offsetFromDefault - other.offsetFromDefault
    }

}

val Octave.dao
    get() = OctaveDao(offsetFromDefault)

val Octave.up
    get() = Octave(offsetFromDefault + 1)

val Octave.down
    get() = Octave(offsetFromDefault - 1)

